internal/controller/testrun_controller.go

```diff
@@ -0,0 +1,405 @@
+// Copyright Axis Communications AB.
+//
+// For a full list of individual contributors, please see the commit history.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/clock"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	etosv1alpha1 "github.com/eiffel-community/etos/api/v1alpha1"
+)
+
+const (
+	typeActive      = "Active"
+	typeSuiteRunner = "SuiteRunner"
+	typeEnvironment = "Environment"
+)
+
+// TestRunReconciler reconciles a TestRun object
+type TestRunReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Clock  clock.WithTicker
+}
+
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=testruns,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=testruns/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=testruns/finalizers,verbs=update
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments,verbs=get;watch;create;delete
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments/status,verbs=get
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/reconcile
+func (r *TestRunReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.V(2).Info("Get testrun", "namespace", req.Namespace, "name", req.Name)
+	testrun := &etosv1alpha1.TestRun{}
+	if err := r.Get(ctx, req.NamespacedName, testrun); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("Testrun not found, exiting", "namespace", req.Namespace, "name", req.Name)
```

Comment:
The error message seems to be the same for both scenarios:
- when the test run doesn't exist
- when there was an error getting the test run
Maybe use different error message.

Reply:
If the testrun does not exist, we cannot do anything. This log message is printed only if the testrun has been deleted.
Outside of the `apierrors.IsNotFound(err)` (line 76) we log the error we got back from `r.Get`